### DISCUSSION TITLE
eq_any for tuples

### DISCUSSION THREAD

**Discussion by SteelAlloy** [diesel-rs/diesel#3577] [Q&A] [answered]
Following https://github.com/diesel-rs/diesel/issues/3222, I see that it is now necessary to filter tuples with a loop and that it is not possible anymore with eq_any.

### Correct
```rust
let mut query = something::table.into_boxed();
for (k1, k2) in inputs {
    query = query.or_filter(something::key_one.eq(k1).and(something::key_two.eq(k2)));
}
let matching = query.load(conn)?;
```

### Invalid
```rust
use diesel::prelude::*;
use diesel::sql_types::Record;

diesel::table! {
    use diesel::sql_types::*;

    something (id) {
        id -> Int4,
        key_one -> Text,
        key_two -> Text,
    }
}

fn main() {
    let inputs: Vec<(String, String)> = vec![("one".to_owned(), "two".to_owned())];
    let conn = todo!();
    let matching = something::table
        .filter(
            (something::key_one, something::key_two)
                .into_sql::<Record<_>>()
                .eq_any(inputs),
        )
        .load(conn);
    println!("Hello, world!");
}

```

It's a performance decision and I understand it. But in some cases I'm not sure it's beneficial all the time.

## Issue

Here is an example of a query that I would like to implement with diesel, but I don't see how to do it without retrieving the result of the first query and injecting it in the second one with a loop as described above.

```sql
SELECT * FROM some_table WHERE (column1, column2) IN (SELECT column1, max(column2) column2 FROM some_table GROUP BY column1);
```

This kind of query can be quite expensive, I would like to do it in one query and not use a buffer variable, especially since the intermediate result can be quite big.

Is there a better way to do it? Thanks!

**Comment by weiznich** [comment, marked as answer]
You can always extend diesels query DSL to include support for otherwise unsupported expressions. See the ["Extending Diesel"](diesel.rs/guides/extending-diesel.html) guide for a conceptual example.
